pkg/util/cgroup/cgroup.go

```diff
@@ -0,0 +1,236 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021
+ *
+ */
+
+package cgroup
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"reflect"
+	"strings"
+
+	"kubevirt.io/kubevirt/pkg/util"
+
+	"github.com/opencontainers/runc/libcontainer/cgroups"
+	"github.com/opencontainers/runc/libcontainer/cgroups/devices"
+	"github.com/opencontainers/runc/libcontainer/cgroups/ebpf"
+	"github.com/opencontainers/runc/libcontainer/cgroups/ebpf/devicefilter"
+	"github.com/opencontainers/runc/libcontainer/cgroups/fscommon"
+	"github.com/opencontainers/runc/libcontainer/configs"
+
+	"golang.org/x/sys/unix"
+)
+
+var (
+	errNoPreciseMatch = errors.New("resulting devices cgroup doesn't precisely match target")
+	errNoMatch        = errors.New("resulting devices cgroup doesn't match target mode")
+)
+
+func BasePath() string {
+	return "/sys/fs/cgroup"
+}
+
+func ControllerPath(controller string) string {
+	if cgroups.IsCgroup2UnifiedMode() {
+		return BasePath()
+	}
+	return filepath.Join(BasePath(), controller)
+}
+
+func Slice(pid int, allowedControllers []string) (controllers []string, slice string, err error) {
+	cgroups, err := os.Open(fmt.Sprintf("/proc/%d/cgroup", pid))
+	if err != nil {
+		return
+	}
+	defer util.CloseIOAndCheckErr(cgroups, nil)
+	scanner := bufio.NewScanner(cgroups)
+	for scanner.Scan() {
+		cgEntry := strings.SplitN(scanner.Text(), ":", 3)
+		// Check if we have a sane cgroup line
+		if len(cgEntry) != 3 {
+			err = fmt.Errorf("Could not extract slice from cgroup line: %s", scanner.Text())
+			return
+		}
+
+		// The entry for cgroup v2 is always in the format "0::$PATH"
+		if cgEntry[0] == "0" && cgEntry[1] == "" && slice == "" {
```

Comment:
Maybe we should spilt the controllers lookup from the slice detection somehow? it looks like this circumvents some of the cgroupv1 checks. The logic is too hard for me to follow now.

Reply:
Agree. There is a nice set of helpers In `runc/libcontainer/cgroups` and I was thinking about refactoring the implementation using those.